module/zfs/zfs_dir.c

```diff
@@ -508,8 +513,51 @@ zfs_unlinked_drain(zfsvfs_t *zfsvfs)
 
 		zp->z_unlinked = B_TRUE;
 		iput(ZTOI(zp));
+		ASSERT3B(zfsvfs->z_unmounted, ==, B_FALSE);
 	}
 	zap_cursor_fini(&zc);
+
+	zfsvfs->z_draining = B_FALSE;
+	zfsvfs->z_drain_task = TASKQID_INVALID;
+}
+
+/*
+ * Sets z_draining then tries to dispatch async unlinked drain.
+ * If that fails executes synchronous unlinked drain.
+ */
+void
+zfs_unlinked_drain(zfsvfs_t *zfsvfs)
+{
+	ASSERT3B(zfsvfs->z_unmounted, ==, B_FALSE);
+	ASSERT3B(zfsvfs->z_draining, ==, B_FALSE);
+
+	zfsvfs->z_draining = B_TRUE;
+	zfsvfs->z_drain_continue = B_TRUE;
+
+	zfsvfs->z_drain_task = taskq_dispatch(
+	    dsl_pool_unlinked_drain_taskq(dmu_objset_pool(zfsvfs->z_os)),
+	    zfs_unlinked_drain_task, zfsvfs, TQ_SLEEP);
+	if (zfsvfs->z_drain_task == TASKQID_INVALID) {
+		zfs_dbgmsg("async zfs_unlinked_drain dispatch failed");
+		zfs_unlinked_drain_task(zfsvfs);
+	}
+}
+
+/*
+ * Wait for the unlinked drain taskq task to stop. This will interrupt the
+ * unlinked set processing if it is in progress.
+ */
+void
+zfs_unlinked_drain_stop_wait(zfsvfs_t *zfsvfs)
+{
+	ASSERT3B(zfsvfs->z_unmounted, ==, B_FALSE);
+
+	if (zfsvfs->z_draining) {
+		zfsvfs->z_drain_continue = B_FALSE;
+		taskq_wait_id(dsl_pool_unlinked_drain_taskq(
+		    dmu_objset_pool(zfsvfs->z_os)), zfsvfs->z_drain_task);
```

Comment:
yeah, looks like a pool-wide taskq is not going to cut it here since if there are lots of tasks in the queue we could be stuck in the back and holding up unmount as a result. I'll see what other options make sense.

As for `iput()` my understanding is in-line with yours - it could take arbitrarily long for large files. There is no getting away from this though - we either wait before we can mount (as done currently) or we have to wait on unmount (like in this patch). The benefit of canceling is that we don't have to wait for ALL the entries in the unlinked set, but only the one that was processing during the unmount.

Reply:
https://github.com/Nexenta/illumos-nexenta/commit/ea0ae4efc65a839131#diff-6d1e73746c8c45ebbc788dd57707c6c2R1888 and https://github.com/Nexenta/illumos-nexenta/commit/5d09257a00d903#diff-26f5af57e103cb3660f924e73c01dc2cR544 are a couple of places where ZoL differs there were more but these are the easiest to find